### DISCUSSION TITLE
Database transaction not working with Lucid relation

### DISCUSSION THREAD

**Discussion by sunxyw** [adonisjs/core#1267] [Help] [answered]
I am trying to create a user and his auth information.
This is my code:
```js
await Database.transaction(async (trx) => {
  const uuid = uuidv4()
  let user = new User()
  user.uuid = uuid
  user.nickname = data.nickname
  user.roleId = 1
  user = user.useTransaction(trx)
  await user.save()

  const auth = new UserAuth()
  auth.fill({
    userUuid: '3x', // throw duplicate key error
    type: data.auth_type,
    identifier: data.identifier,
    credential: data.credential,
  })
  auth.useTransaction(trx)
  await auth.save()
})
```
My `User` model: https://pastebin.com/gDDLQHxM
My `UserAuth` model: https://pastebin.com/4BQgrTch

The error happened as expected, but the user record didn't rollback.

Is there any mistake in my code? Thanks for your help.

**Comment by thetutlage** [comment, marked as answer]
Just by looking at the code, it should work. Couple of questions?

- Which version of `@adonisjs/lucid` you are using?
- Can you turn on [query debugging](https://preview.adonisjs.com/guides/database/debugging-queries#debugging-queries) and share the console output?

**Reply by sunxyw** [reply]
Sorry, I forgot to provide my Lucid version.

I am using `"@adonisjs/lucid": "^8.2.2"`.

sql log:
```
"mysql" User (1.57 ms) insert into `users` (`created_at`, `nickname`, `role_id`, `updated_at`, `uuid`) values (?, ?, ?, ?, ?) [
  '2020-07-19 15:14:59',
  'Jimmy',
  1,
  '2020-07-19 15:14:59',
  'b4fe7abb-7b5f-4614-a58d-60d2d9ed396a'
]


"mysql" UserAuth (5.43 ms) insert into `user_auths` (`created_at`, `credential`, `identifier`, `type`, `updated_at`, `user_uuid`) values (?, ?, ?, ?, ?, ?) [
  '2020-07-19 15:14:59',
  '123456',
  '.@yahoo.com',
  'email',
  '2020-07-19 15:14:59',
  '3x'
]

[1595142899592] ERROR (mist/2658 on sunxyw-laptop): insert into `user_auths` (`created_at`, `credential`, `identifier`, `type`, `updated_at`, `user_uuid`) values ('2020-07-19 15:14:59', '123456', '.@yahoo.com', 'email', '2020-07-19 15:14:59', '3x') - ER_DUP_ENTRY: Duplicate entry '3x-email' for key 'user_auths_user_uuid_type_unique'
    x-request-id: "ckcsqwafk00001u4c58rnc88x"
```

**Reply by thetutlage** [reply]
If I look at the above logs, you have two calls to insert into `user_auths`, which shouldn't be the case as per your code.

Also, I am not sure, if you shared the exact code you are running in your app. It will be nice, if you can create a project with the minimal required code to reproduce this issue

**Reply by sunxyw** [reply]
I created a project to reproduce this issue.
Please check [https://github.com/sunxyw/adonisjs-transaction](https://github.com/sunxyw/adonisjs-transaction)

**Reply by thetutlage** [reply]
Just tried. Works like a charm for me. https://res.cloudinary.com/adonis-js/video/upload/v1595147217/transactions-check_l6gldv.mp4

**Reply by sunxyw** [reply]
However, It doesn't work on my computer, https://res.cloudinary.com/sunxywcloud/image/upload/v1595147706/fe8af96d-1b42-4096-b471-2fadf489f205_ge7daz.gif.
Maybe I should check my MySQL setting.

Anyway, thanks for your help.

**Reply by sunxyw** [reply]
Well.....I guess I find the point.
It should be a bug of HeidiSQL (Database Manage Tool).

The record is rolled back actually, but the overview in HeidiSQL doesn't change. It seems that it used Cache.
That's why [this](#discussioncomment-39177) happened.

When I enter the detail page to confirm, everything works fine.

**Reply by thetutlage** [reply]
Glad, you found it :)

**Comment by sunxyw** [comment]
Btw, I find something interesting.

When inserted about 5-6 records, it will be rollback.

For example, now have 35 `users` records, after I send 5-6 requests to API(create user), It will be increased to 40.
After one more request, It will roll back to 35.

I have no idea why it will happen.

~~I know my English is quite poor, hope you can understand what I mean.~~

**Comment by popovicieduard** [comment]
I have the same issue, I cloned the repo provided by @sunxyw , and the users are insterted, while the user auths are not.

postman call
https://prnt.sc/toc9k3

insterted users:
https://prnt.sc/toc9sl

user auths:
https://prnt.sc/toc9ze

db log query:
```
"mysql" User (4.19 ms) insert into `users` (`created_at`, `nickname`, `role_id`, `updated_at`, `uuid`) values (?, ?, ?, ?, ?) [
  '2020-07-25 14:59:00',
  'test',
  1,
  '2020-07-25 14:59:00',
  '3efd924f-9d10-4646-9bdc-4639c9980f9e'
]
"mysql" UserAuth (31 ms) insert into `user_auths` (`created_at`, `credential`, `identifier`, `type`, `updated_at`, `user_uuid`) values (?, ?, ?, ?, ?, ?) [
  '2020-07-25 14:59:00',
  'test',
  'foo',
  'login',
  '2020-07-25 14:59:00',
  '3x'
]
```

**Reply by sunxyw** [reply]
It should be normal that `user_auths` is not inserted because the `user_uuid` value of `user_auths` is set to a fixed value `3x`, which will trigger the duplicate unique index error.

You may change `3x` to `uuid` to solve it in https://github.com/sunxyw/adonisjs-transaction/blob/master/start/routes.ts#L52

**Reply by popovicieduard** [reply]
The expected behaviour would be that the transaction rolls back and the user is not insterted, right?, like @thetutlage showed: https://res.cloudinary.com/adonis-js/video/upload/v1595147217/transactions-check_l6gldv.mp4

**Reply by sunxyw** [reply]
Yes, both `user_auths` and `users` should be rollback.

**Reply by popovicieduard** [reply]
Right, but for some reason, users are insterted and user auths are not, users are insterted even if it throws duplicate error

**Reply by sunxyw** [reply]
oh, I got your point.
However, I have no idea about this issue.
Because my question is actually just a mistake because I forgot to check the detailed data page.

Therefore, I think it will be better to create a new thread to solve the issue, because this thread is already marked as resolved (And it seems cant be reopen)

**Reply by popovicieduard** [reply]
Found out the issue, Transactions in mysql only work with some storage type: https://stackoverflow.com/questions/42436932/transactions-not-working-for-my-mysql-db .
